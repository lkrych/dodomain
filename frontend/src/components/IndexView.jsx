import React, {Component} from 'react';
import Typography from '@material-ui/core/Typography';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import TablePagination from '@material-ui/core/TablePagination';
import Paper from '@material-ui/core/Paper';

class IndexView extends Component {
  constructor(props){
    super(props);
    this.state = {
      page: 1,
      search_query: "",
      desc: true,
      order_by: "id",
      itemsPerPage: 50,

    };
    this.handleChangePage = this.handleChangePage.bind(this);
  }

  handleChangePage (event, page){
    //don't pass in unneccessary parameters
    let submitState = Object.assign({}, this.state, {page: page+1});
    delete submitState['itemsPerPage'];
    event.preventDefault();
    this.setState({ page: page + 1 }, () => {
      this.props.fetchDomains(submitState);
    });
  }

  componentDidMount(){
    //don't pass in unneccessary parameters
    let submitState = Object.assign({}, this.state);
    delete submitState['itemsPerPage'];
    this.props.fetchDomains(submitState);
  }

  render(){
    const { classes } = this.props;
    let domains = [];
    if (this.props.domains.domains) {
      domains = this.props.domains.domains;
    }
    let textErrors = <Typography color="error" variant="body2" id="errors" >{this.props.errors}</Typography>;

    return (
      <Paper className={classes.root}>
        <Typography variant="headline" gutterBottom>
          List of Valid Domains
        </Typography>
        {textErrors}
        <Table className={classes.table}>
          <TableHead>
            <TableRow>
              <TableCell>Name</TableCell>
              <TableCell>Description</TableCell>
              <TableCell >Created At</TableCell>
            </TableRow>
          </TableHead>
          <TableBody className="domain-list">
            {domains.map((n, idx) => {
              return (
                <TableRow key={idx}>
                  <TableCell component="th" scope="row">
                    {n.name}
                  </TableCell>
                  <TableCell>{n.description}</TableCell>
                  <TableCell>{new Date(n.created_at).toLocaleString()}</TableCell>
                </TableRow>
              );
            })}
          </TableBody>
        </Table>
        <TablePagination
            component="div"
            count={this.props.domains.pagination.totalItems}
            rowsPerPage={this.state.itemsPerPage}
            page={this.state.page - 1} //uses zero-based index
            rowsPerPageOptions={[50]}
            backIconButtonProps={{
              'aria-label': 'Previous Page',
              'id': 'back-button'
            }}
            nextIconButtonProps={{
              'aria-label': 'Next Page',
              'id': 'forward-button'
            }}
            onChangePage={this.handleChangePage}
          />
      </Paper>
    );
  }
}

export default IndexView;