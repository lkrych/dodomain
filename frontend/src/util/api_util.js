const buildRequestHeaders = () => {
  const headers = new Headers();
  headers.append('Content-Type', 'application/json');
  headers.append('Authorization', `Bearer ${sessionStorage.token}`);
  return headers;
};

// buildurl is used by the fetchResources function to create the queryString
// that will determine what resources are fetched from the back end

export const buildUrl = (resource, sortObject) => {
  let url = '/' + resource + '?';
  let sortParams = Object.keys(sortObject);
  sortParams.forEach((key, idx) => {
    if (sortObject[key] === null) { // catch bad data
      url = url + key + '=';
      url += '&';
    } else {
      if (typeof sortObject[key] === "object")
        url = url + key + '=' + JSON.stringify(sortObject[key]);
      else
        url = url + key + '=' + sortObject[key];
        url += '&';
    }
  });

  return url;
};

export const signUp = (user) => {
  const headers = buildRequestHeaders();
  return fetch('/signup', {
    method: 'POST',
    credentials: 'include',
    headers: headers,
    body: JSON.stringify({user})
    }).then(response => {
     return response.json();
    }).then(json => {
    if(json.auth_token){
      sessionStorage.setItem('token', json.auth_token);
    }
    return json;
  });
};

export const login = (user) => {
  const headers = buildRequestHeaders();
  return fetch('/auth', {
    method: 'POST',
    credentials: 'include',
    headers: headers,
    body: JSON.stringify({user})
    }).then(response => {
     return response.json();
    }).then(json => {
    if(json.auth_token){
      sessionStorage.setItem('token', json.auth_token);
    }
    return json;
  });
};

export const fetchDomains = (sortState) => {
  const headers = buildRequestHeaders();
  let url = buildUrl('domains', sortState);
  return fetch(`${url}`, {
    method: 'GET',
    credentials: 'include',
    headers: headers
    }).then(response => {
     return response.json();
  });
};

export const submitDomain = (domain) => {
  const headers = buildRequestHeaders();
  return fetch('/domains', {
    method: 'POST',
    credentials: 'include',
    headers: headers,
    body: JSON.stringify(domain)
    }).then(response => {
     return response.json();
  });
};