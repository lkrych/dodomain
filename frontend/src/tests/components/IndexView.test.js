import React from 'react';
import { shallow } from 'enzyme';

import IndexView from '../../components/IndexView';
import { fetchDomains } from '../../util/api_util';

describe('IndexView', () => {
  const mockLifeCycle = jest.fn();
  const indexView = shallow(<IndexView domains={{domains: [], pagination: {}}} fetchDomains={mockLifeCycle} />);

  it('renders correctly', () => {
    expect(indexView).toMatchSnapshot();
  });

  it('should call fetchDomains when it loads', () => {
    expect(mockLifeCycle.mock.calls.length).toBeGreaterThanOrEqual(1);
  });

  describe('when returned domains from backend', () => {
    let populatedIndexView = shallow(<IndexView domains={{domains: [{name: "test.com"}, {name: "testing.com"}], pagination: {}}} fetchDomains={mockLifeCycle} />);
    it('displays them in a list', () => {
      expect(populatedIndexView.find('.domain-list').children().length).toBeGreaterThanOrEqual(2);
    });
  });
});