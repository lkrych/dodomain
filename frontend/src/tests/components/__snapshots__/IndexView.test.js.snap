// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndexView renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(IndexView)
    domains={
        Object {
            "domains": Array [],
            "pagination": Object {},
          }
    }
    fetchDomains={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "classes": Object {
        "root": "IndexView-root-1",
        "table": "IndexView-table-2",
      },
      "domains": Object {
        "domains": Array [],
        "pagination": Object {},
      },
      "fetchDomains": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "classes": Object {
          "root": "IndexView-root-1",
          "table": "IndexView-table-2",
        },
        "domains": Object {
          "domains": Array [],
          "pagination": Object {},
        },
        "fetchDomains": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
